ddsc/core/remotestore.py

```diff
@@ -0,0 +1,323 @@
+        return RemoteUser(response)
+
+    def fetch_all_users(self):
+        """
+        Retrieves all users from data service.
+        :return: [RemoteUser] list of all users we downloaded
+        """
+        page = 1
+        per_page = FETCH_ALL_USERS_PAGE_SIZE
+        users = []
+        while True:
+            result = self.data_service.get_users_by_page_and_offset(page, per_page)
+            user_list_json = result.json()
+            for user_json in user_list_json['results']:
+                users.append(RemoteUser(user_json))
+            total_pages = int(result.headers["x-total-pages"])
+            result_page = int(result.headers["x-page"])
+            if result_page == total_pages:
+                break;
+            page += 1
+        return users
+
+    def set_user_project_permission(self, project, user, auth_role):
+        """
+        Update remote store for user giving auth_role permissions on project.
+        :param project: RemoteProject project to give permissions to
+        :param user: RemoteUser user who we are giving permissions to
+        :param auth_role: str type of authorization to give user(project_admin)
+        """
+        self.data_service.set_user_project_permission(project.id, user.id, auth_role)
+
+    def revoke_user_project_permission(self, project, user):
+        """
+        Update remote store for user removing auth_role permissions on project.
+        :param project: RemoteProject project to remove permissions from
+        :param user: RemoteUser user who we are removing permissions from
+        """
+        # Server errors out with 500 if a user isn't found.
+        try:
+            resp = self.data_service.get_user_project_permission(project.id, user.id)
+            self.data_service.revoke_user_project_permission(project.id, user.id)
+        except DataServiceError as e:
+            if e.status_code != 404:
+                raise
+
+    def download_file(self, remote_file, path, watcher):
+        """
+        Download a remote file associated with the remote uuid(file_id) into local path.
+        :param remote_file: RemoteFile file to retrieve
+        :param path: str file system path to save the contents to.
+        :param watcher: object implementing send_item(item, increment_amt) that updates UI
+        """
+        url_json = self.data_service.get_file_url(remote_file.id).json()
+        http_verb = url_json['http_verb']
+        host = url_json['host']
+        url = url_json['url']
+        http_headers = url_json['http_headers']
+        response = self.data_service.receive_external(http_verb, host, url, http_headers)
+        with open(path, 'wb') as f:
+            for chunk in response.iter_content(chunk_size=DOWNLOAD_FILE_CHUNK_SIZE):
+                if chunk:  # filter out keep-alive new chunks
+                    f.write(chunk)
+                    watcher.sending_item(remote_file, increment_amt=len(chunk))
```

Comment:
Was watcher repurposed from the upload side? `sending_item` is confusing for download. maybe `transferring_item` ?